### ISSUE TITLE
Make 1.0.6 a release

### ISSUE THREAD

**Issue by erlanger** [olsky/swi-mqtt-pack#2]
Hi Olsky,

   Thank you very much for the mqtt pack, it is very useful. Could you please make 1.0.6 a release? The current swi-prolog add-on list has the older 1.0.5 version, which has a serious bug preventing subscription of several topics.

Thanks!

**Comment by sprior** [comment]
As far as I know the Olsky version of this library is not being maintained anymore, I've got a version which works for me in my fork sprior/swi-mqtt-pack, though I've never done a "release" - it's source code only.  I'm not sure what bug you're referring to, is this a bug in the upstream mqtt library?

**Comment by erlanger** [comment]
@sprior,

Not a mosquitto bug, it is a bug in  olsky/swi-mqtt-pack which is solved by committ [db890bb5e9645ef5fff1dc9804663fa4bffe42e2](https://github.com/olsky/swi-mqtt-pack/commit/db890bb5e9645ef5fff1dc9804663fa4bffe42e2#diff-29f662de5188e4a88f8c5fc83e0b4692L730) in this repository. 

The problem is that `c_mqtt_sub` is calling `mosquitto_reconnect`, which clears all previous subscriptions.

This happens  because the `clean session` flag is set to true when `mosquitto_new` is called. The net effect that one can subscribe only to one topic per connection. Olsky already updated the version to 1.0.6 which fixes this bug, all he neeeds to do is make a github release so that the swi-prolog website picks it up.

If @olsky  is not maintaining it, perhaps you could offer to maintain it by sending an email to Jan and move the repository to yours. That would be a great contribution to  swi-prolog!

Thanks!

**Comment by sprior** [comment]
I just put out a 1.0.7 release (it has all the olsky commits plus some changes I made which add examples and made it work for me.  Give it a try and if you agree it works I'll contact Jan.  One change I made is that I do not check binaries into git  and only the library for Prolog is included in the release - the mosquitto library should be installed via its own package.

**Comment by erlanger** [comment]
@sprior,

Good work, I am glad you are planning to contribute to Swi prolog! 

Your version does not compile for me, because you rely on the `PROLOG_HOME` and the `PROLOG_VER` environment variables. 

SWI prolog provides the proper `CFLAGS` when you run  `pack_rebuild(mqtt)`, so you don't need to use `PROLOG_HOME` or `PROLOG_VER`. In order for this to work you need to make the following change:
```sh
CFLAGS=$(COFLAGS) $(CWFLAGS) $(CMFLAGS) $(CIFLAGS) $(PKGCFLAGS) $(DEFS) -I$(PROLOG_HOME)/lib/swipl-$(PROLOG_VER)/include
```
into
```sh
CFLAGS+=$(COFLAGS) $(CWFLAGS) $(CMFLAGS) $(CIFLAGS) $(PKGCFLAGS) $(DEFS)
```

The `+=` will ensure using the SWI prolog provided environment during `pack_rebuild(mqtt)`.

 A lot more could be cleaned up in the Makefile,  but with this change everything compiles and works for me.

If you want to build from the shell command line (after the `+=` change) you need to source the `buildenv.sh` file which is automatically generated by SWI prolog during `pack_rebuild(mqtt)`.

I am glad you are contributing to SWI prolog!

P.S. For the future I would suggest to split cosmetic changes and code changes into different commits, to keep the commit size small and to make it easy to review.

**Comment by sprior** [comment]
I've actually been building/deploying my SWI-Prolog home automation code using Docker.  Here's the Dockerfile I've been using to create a SWI-Prolog installation with MQTT included (yes I know the resulting container could be made more minimal with a two stage build, but haven't been motivated to go there yet).

`FROM ubuntu:xenial
MAINTAINER Steve Prior <sprior@geekster.com>
ENV TZ "EST5EDT"
ENV ENVIRONMENT_REFRESH 2018-09-25

LABEL Description="SWI Prolog devel binary install with mqtt"

RUN apt-get update && apt-get install -y git \
    libmosquitto-dev \
    gcc \
    software-properties-common \
    make && \
    apt-add-repository ppa:swi-prolog/devel && \
    apt-get update && \
    apt-get install -y swi-prolog && \
    git clone https://github.com/sprior/swi-mqtt-pack.git && \
    cd swi-mqtt-pack && \
    mv Makefile Makefile.bak && \
    sed 's#$(PROLOG_HOME)/lib/swipl-$(PROLOG_VER)/include#/usr/lib/swi-prolog/include#' Makefile.bak > Makefile && \
    make && \
    ln -s /swi-mqtt-pack/lib/x86_64-linux/mqtt.so /usr/lib/swi-prolog/lib/amd64 && \
    ln -s /swi-mqtt-pack/prolog/mqtt.pl /usr/lib/swi-prolog/library`

Do you have any suggestions on how this would need to be changed to include your suggestions?  I'm not familiar with the pack_rebuild command line.

I'm sure a lot more can be cleaned up in both the Makefile and the code.  Once I saw the original module I hadn't planned to get into working on it and only did so by necessity and kept my changes as small as I could tolerate.

**Comment by olsky** [comment]
Hi all,
I've pushed the release 1.0.6. In case you have edits/commits, please use "pull request", I will merge it.

Thanks and cheers!
Oleh

**Comment by erlanger** [comment]
Hi @olsky,

  Thanks for reviving your package!

  The swi prolog website has not picked up your [update](http://www.swi-prolog.org/pack/list?p=mqtt).
  The reason is that your `pack.pl` file is not setup for automatic updates.
You need to change this line in `pack.pl`:
`download('https://github.com/olsky/swi-mqtt-pack/releases/download/v1.0.6/mqtt-1.0.6
.zip').`

into this:

`download( 'https://github.com/olsky/swi-mqtt-pack/releases/*.zip' ).`

Otherwise you need to contact Jan to make the update.

As mentioned in the [documentation](http://www.swi-prolog.org/howto/Pack.html).

  I will submit a pull request as soon as I can, to make it compile using the swi prolog environment and to use the mosquitto shared library already installed in the system.

Hopefully @sprior will submit a patch with his new feature supporting binary mqtt messages.

**Comment by sprior** [comment]
@erlanger For the moment I could really use your help with the Makefile/build changes you mentioned, I'm trying to figure out exactly what needs to be changed in the Dockerrfile I posted above.  I looked into pack_rebuild, but there seems to be no useful documentation on how to use it.  It's also possible that I'd need to name my version of the mqtt pack something else to not collide with the package management system or I need to put the build directory in a certain place - I'm really not sure.  If you would be willing to contact me off of this issue to help work out those things (since I'd imagine olsky would consider them offtopic) I'd be grateful.

Having gone without any activity for so long I really thought the original branch had been abandoned and had to go through a lot of effort to come up with a set of working examples.  I've been using git for a few years in my day job, but not as familiar with github - this is the only core code that's part of my home automation system that's hosted on github instead of my private git server.  That's maybe why I wasn't as careful to keep function/style commits distinct from each other.

**Comment by erlanger** [comment]
Hi @sprior,

  I think your examples are really useful, and also the new feature to support binary payload. I would suggest you submit two separate pull requests to @olsky , one with the binary payload feature and another with the examples. 

I will submit a pull request with simple changes to the makefile which will allow it to work the way swi prolog packs are supposed to work. You can then pick up the changes from that pull request. 

`pack_rebuild()` is used within swi-prolog. So the normal process to rebuild a prolog pack would be something like this:

```sh
$ swipl
Welcome to SWI-Prolog (threaded, 64 bits, version 7.7.19)
SWI-Prolog comes with ABSOLUTELY NO WARRANTY. This is free software.
Please run ?- license. for legal details.

For online help and background, visit http://www.swi-prolog.org
For built-in help, use ?- help(Topic). or ?- apropos(Word).

1 ?- pack_rebuild(mqtt).
% rm -f c/mqtt.o
% rm -f lib/x86_64-linux/mqtt.so
% gcc -pthread -fPIC -D_FORTIFY_SOURCE=2 -I"/usr/lib/swipl-7.7.19/include" -O2 -g -Wall -fno-strict-aliasing -pthread -fPIC -std=c99 -I. -Iext/include     -c -o c/mqtt.o c/mq
tt.c
% mkdir -p lib/x86_64-linux
% gcc -shared -rdynamic -O2 -g -Lext -pthread -Wl,-rpath=ext -Lext -lrt -lm -lpthread -o lib/x86_64-linux/mqtt.so  c/mqtt.o -lmosquitto
% make: Nothing to be done for 'check'.
% make: Nothing to be done for 'install'.
true.

2 ?-
```

This can be done from the command line by running:
```sh
$ swipl -g 'pack_rebuild(mqtt)' -t 'halt(0)'
% rm -f c/mqtt.o
% rm -f lib/x86_64-linux/mqtt.so
% gcc -pthread -fPIC -D_FORTIFY_SOURCE=2 -I"/usr/lib/swipl-7.7.19/include" -O2 -g -Wall -fno-strict-aliasing -pthread -fPIC -std=c99 -I. -Iext/include     -c -o c/mqtt.o c/mq
tt.c
% mkdir -p lib/x86_64-linux
% gcc -shared -rdynamic -O2 -g -Lext -pthread -Wl,-rpath=ext -Lext -lrt -lm -lpthread -o lib/x86_64-linux/mqtt.so  c/mqtt.o -lmosquitto
% make: Nothing to be done for 'check'.
% make: Nothing to be done for 'install'.
$
```

The above is the output with the corrected `Makefile` from pull request #3 .
